### DISCUSSION TITLE
2023-05-31 Meeting

### DISCUSSION THREAD

**Discussion by thomas-fossati** [ietf-rats-wg/draft-ietf-rats-corim#105] [General]
# Agenda

* Discuss https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/95 and [composition musing](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/wiki/Composition-musings)

* Discuss (and resolve if possible) open PRs:
  * #94
  * #104

**Comment by thomas-fossati** [comment]
# Minutes

* Potential gap in the RATS architecture related to composition

## Actions

- [x] @nedmsmith to write up and socialise with the wider RATS WG.

**Reply by nedmsmith** [reply]
Summary of issue:
Today verifiers depend on Evidence to prescribe how measurements are grouped. For example, a DICE certificate contains DiceTcbInfo, DiceTcbInfoSeq, or other variations of evidence. Each DICE certificate is an instance of Evidence. If using EAT/CWT, each token is an instance of Evidence. The verifier assumes the "device" has been attested based on the received evidence(s). If an alias certificate / token is dropped, the verifier might not realize the device is only partially attested.

CoRIM has some techniques to ensure the device is attested completely. Domain-membership and domain-dependency-triples can be used to describe groupings of environments (and their measurements via reference-value-triples, and identity-triples) and other groupings (domains) that depend on another grouping (e.g., the DICE layering).  For these triples to ensure there is complete device coverage, every component is represented as a TE that is a member of a domain that has a trust dependency on one or more attestation keys used to sign evidence.

The RATS Architecture didn't define measurement grouping terminology other than to describe `composite device` as an entity composed of multiple Attesters (aka lead Attester and sub-Attesters). It implies that an 'environment' (TE/AE) could have multiple measurements, but doesn't try to define a grouping of measurements (claims) that belong to a particular environment. It suggests that an AE could collect measurements from multiple TEs, which implies a grouping of TEs. It shows an AE within a TE, implying a hierarchy of both TEs and AEs is possible. 

However, Evidence appears to be the only expression that contains (actual) grouping / nesting. The AEs / lead Attester(s) generate and group Evidence. Nevertheless, it is possible / reasonable that each AE / lead Attester doesn't try to represent grouping / nesting semantics as each signed Evidence block could be serialized in some arbitrary order for conveyance. 

Grouping of measurements based on the token, cert, etc., as a way of specifying device composition is merely conjecture.  

The RATS Architecture suggests that a Verifier is aware of device composition and is responsible for verifying the device is fully attested, but doesn't describe how that is achieved. It seems to suggest that a local Verifier could be co-located with a lead Attester to ensure attestation Evidence completely covers the device. Use of an Endorsement or Reference Value to describe device composition is not described, though may be implied given the Verifier is supposed to know how to do it.

**Reply by thomas-fossati** [reply]
Thanks for the write up @nedmsmith!

My reading of the architecture is that it provides vocabulary for all the building blocks that can be used in composition patterns.

In particular, I don’t think there is a real justification to mint the term “domain” in CoRIM, since "composite device" and, even better, "attester" already exist and can be readily used in its stead as grouping constructs for "environments".

I agree with your analysis that RFC9334 does not provide terminology for precise description of the topological arrangement of a composite device in terms of its component attesters and their internal relationship.  However, I don't think the architecture is tasked to solve that specific problem.  As long as it provides the logical framework where all the needed entities can be named, it's the task of the info models that are based on the architecture to create the semantic relationships they need (e.g., via CoRIM triples).

**Reply by henkbirkholz** [reply]
> If an alias certificate / token is dropped, the verifier might not realize the device is only partially attested.
Hm... really? Only if a required set of RV or some composition reference would not be met, right? Maybe I am missing something obvious.

From my PoV, it is utterly important for the Verifier to always have a consistent awareness about which Attesting Environment is intended to provide Evidence about which Target Environment in the scope of an Attester! On the one hand, that is a hard requirement that stems from [Composite Devices](https://www.rfc-editor.org/rfc/rfc9334.html#name-composite-device), I think.

On the other hand, a Verifier must always have a consistent awareness about which RV and endorsed values are required to be considered complete&ready for evidence appraisal wrt an identified Target Environment. Following this thought, wouldn't a domain not be equivalent to a Target Environment? If a domain, for example, is a group of more than one Target Environment, a grouping of RV and endorsed values (e.g., by domain) could make sense, I think. But I am not certain that grouping by Target Environment alone would be less expressive than a thing to express groups of Target Environments (if CoRIM informs the verifier from which Attesting Environment to expect Evidence about certain Target Environments). Is there a counter-example? Am I missing the point?
